runtime/compiler/runtime/JITServerROMClassHash.cpp

```diff
@@ -52,3 +82,23 @@ JITServerROMClassHash::toString(char *buffer, size_t size) const
       s += sprintf(s, "%02x", ((uint8_t *)_data)[i]);
    return buffer;
    }
+
+
+volatile bool JITServerROMClassHash::_cachedObjectArrayHash = false;
+JITServerROMClassHash JITServerROMClassHash::_objectArrayHash;
+
+const JITServerROMClassHash &
+JITServerROMClassHash::getObjectArrayHash(const J9ROMClass *objectArrayROMClass, TR_Memory &trMemory, TR_J9VMBase *fej9)
+   {
+   if (_cachedObjectArrayHash)
+      {
+      TR_ASSERT(_objectArrayHash == JITServerROMClassHash(objectArrayROMClass, trMemory, fej9),
+                "Mismatching cached object array ROMClass hash");
+      return _objectArrayHash;
+      }
+
+   _objectArrayHash.init(objectArrayROMClass, trMemory, fej9);
```

Comment:
This could result in several threads executing in parallel:
`success = OEVP_DigestFinal_ex(ctx, (uint8_t *)_data, &hashSize);`

If those threads executed this call sequentially, the end result would be the same regardless of the order. However, because we don't know the implementation of `DigestFinal` it may be test to protect the code with some mutex.

Reply:
For future reference: if in the future we ever need to call this function in a truly multi-threaded context (not inside a critical section like it is now), a solution that avoids locks is to compute the hash into a local stack buffer first, and then copy it word-by-word (which should be atomic on our target architectures) into the shared static field.